File: algorithms/Strings/Medium/WordBreak/VSDWordBreak.py

def wordBreak(string, words):

    d=[False]*len(string)

    for i in range(len(string)):
        for w in words:

            if w == string[i-len(w)+1:i+1] and (d[i-len(w)] or i-len(w) == -1):
                d[i]=True
            
    print(d)
    return d[-1]

if __name__=="__main__":
    string="applepenapple"
    words=["apple","pen"]
    print(wordBreak(string,words))

File: algorithms/Strings/Medium/WordBreak/VSDWordBreak.cpp

#include<bits/stdc++.h>

using namespace std;



bool wordBreak(std::string s,std::vector<std::string>& wordDict){
    std::set<std::string> word_set(wordDict.begin(),wordDict.end());
    int n=s.size();
    std::vector<bool>dp(n+1,0);
    dp[0]=1;
    for(int i=0;i<n;i++){
        if(!dp[i]) continue;
        for(int j=i+1;j<=n;j++){
            if(word_set.count(s.substr(i,j-i)))
                dp[j]=1;
        }
    }
    return dp[n];
}

int main(){
    std::string s = "applepenapple";
    std::vector<std::string> words{ "apple", "pen" };
    std::cout<<wordBreak(s,words);
    return 0;
}